### ISSUE TITLE
update_cache() can take a long time

### ISSUE THREAD

**Issue by glommer** [scylladb/scylla#1570]
During the default cassandra-stress workload against a single core, row_cache::update can take a long time between yields. Instrumenting the main while loop manually, we can see that this can be as high as 38ms.

Moving on, we can try finding the problematic points. The first question we must ask ourselves is: is this happening because we are executing the loop many times, and should yield more often, or is any one operation alone exploding the cost ?

To find this out, I have added time measurements inside the loop itself, and took the maximum between all loop iterations. With 100 % correlation, runs that took a long time has one big spiky run that bump the time upwards. Example output:

```
ROW CACHE RAN FOR 37 msec, MAX 37999 usec
ROW CACHE RAN FOR 7 msec, MAX 7999 usec
```

I have then separated one iteration into three parts and timed then independently

Part 1: `auto cache_i = _partitions.lower_bound(mem_e.key(), cmp);`
Part 2: `if (cache_i != _partitions.end()... else { ... }`
Part 3: `i = m.partitions.erase(i); current_allocator().destroy(&mem_e);`

Out of 28 "bad loops", 2 of them were in part 2. One taking 2ms, and another 3ms. The remaining 26 were in part 1, with biggest offender at 32ms, but routinely over 10ms.

That is calculating the lower_bound on a boost intrusive set.

**Comment by tgrabiec** [comment]
Are you sure those latencies are not due to scylla being preempted by another process? 30ms seems excessive for a single lookup.

**Comment by glommer** [comment]
On Wed, Aug 10, 2016 at 1:37 PM, Tomasz Grabiec <notifications@github.com>
wrote:

> Are you sure those latencies are not due to scylla being preempted by
> another process? 30ms seems excessive for a single lookup.
>
It could be, of course - although it's suspicious to have the preemptions
always happening at the same point.

But I don't have anything else running in the machine. Could still be a
kernel thread, though. However, I am not concerned too much about this one
run that took 30ms. I am a lot more concerned about all the other ones that
took anywhere from 3 to 20 ms.

This shouldn't take even 1 ms.

**Comment by tgrabiec** [comment]
If preemptions are happening, they could be responsible for shorter latencies as well.

You could try to profile sched:sched_switch perf events to see if non voluntary preemptions are happening.

Is Scylla fully loaded (reactor.load)? Are you using sleep mode?

**Comment by glommer** [comment]
On Wed, Aug 10, 2016 at 2:33 PM, Tomasz Grabiec <notifications@github.com>
wrote:

> If preemptions are happening, they could be responsible for shorter
> latencies as well.
>

This is very unlikely. When we write a memtable to disk, our ops/sec gets
almost halved. And this is not CPU nor IO bound (reactor load is at around
60 % tops). Those drops correlate perfectly with tasks taking way too long.
The tasks that take too long, take more or less exactly the amount of time
this is taking.



> You could try to profile sched:sched_switch perf events to see if non
> voluntary preemptions are happening.
>

They are not. I am on sleep mode (answering your question below), and most
of what I see is just the move to idle. The rest is just kernel threads
that seems to me would be there anyway. (maybe I am missing something there
- would be happy to provide the full output of perf report -T)

> Is Scylla fully loaded (reactor.load)? Are you using sleep mode?
>

I am using sleep mode, yes. Just tried poll mode and I don't see any
significant difference. The max I've seen on poll mode is 29 msecs, but
again, I am not too much concerned with the peak.

I will investigate this more later.

**Comment by glommer** [comment]
For the reference, there are around 400k keys in the cache. If I modify the code to always flush the cache before calling update cache, then I no longer see update_cache taking any relevant time. I still see tasks taking a lot of time, but the problem is moved somewhere else.

But in the interest of going "one problem at a time", @tgrabiec , do you think it could be that manipulating that set becomes that expensive when there are that many entries ? 400k doesn't seem like that many , but it should be easy to isolate it if this is the issue. I will give it a try tomorrow

**Comment by avikivity** [comment]
400k (assuming for that shard) means < 20 compares if the tree is perfectly balanced, perhaps 40 if not.

Even if each compare takes 10 usec, we get 400 usec = 0.4 ms.

The only explanation I can see is that tree balancing is not working.

Note that we insert items in an already sorted order, perhaps that is triggering a bug in the boost tree code.

You can try switching to avl_set or one of the other similar sets to see if it gets us away from this bad behavior.

Another thing you can try is to instrument the compare function to increment some global, then sample the global before and after lower_bound to see how expensive it was.

**Comment by tgrabiec** [comment]
2016-08-11 1:15 GMT+02:00 Glauber Costa <notifications@github.com>:

> On Wed, Aug 10, 2016 at 2:33 PM, Tomasz Grabiec <notifications@github.com>
> wrote:
>
> > If preemptions are happening, they could be responsible for shorter
> > latencies as well.
> >
>
> This is very unlikely. When we write a memtable to disk, our ops/sec gets
> almost halved. And this is not CPU nor IO bound (reactor load is at around
> 60 % tops). Those drops correlate perfectly with tasks taking way too long.
> The tasks that take too long, take more or less exactly the amount of time
> this is taking.
>
>
>
> > You could try to profile sched:sched_switch perf events to see if non
> > voluntary preemptions are happening.
> >
>
> They are not. I am on sleep mode (answering your question below), and most
> of what I see is just the move to idle. The rest is just kernel threads
> that seems to me would be there anyway. (maybe I am missing something there
> - would be happy to provide the full output of perf report -T)
>

Ok, please send `perf report --stdio` so that we're double sure. I'm
assuming that was taken when you observed high latencies.



>
> > Is Scylla fully loaded (reactor.load)? Are you using sleep mode?
> >
>
> I am using sleep mode, yes. Just tried poll mode and I don't see any
> significant difference. The max I've seen on poll mode is 29 msecs, but
> again, I am not too much concerned with the peak.
>
> I will investigate this more later.
>

**Comment by glommer** [comment]
@tgrabiec I have monitored for some 2 seconds before this shows up, and then for some 2 seconds more later.

I have attached the result of perf report -T. I don't have the original perf data anymore, as I have run other things that have overwritten the file. In particular, I have run a flamegraph. My belief is that such a hog should show up somewhere, even if it only happens for a small time. Yet, I see no trace of it.

This leads me to believe that maybe this is taking some kind of internal lock and blocking? In the past, what did you guys use to find sources of userspace locking ? (note that I have no reason to say for sure this is the case, just a wild theory)


[sched.txt](https://github.com/scylladb/scylla/files/413348/sched.txt)

**Comment by glommer** [comment]
@avikivity I will try this. What really puzzles me, though, is that I am not forcing a cleanup of the cache before every update. So this doesn't happen anymore, *but*, the big spikes keep happening. They just move somewhere else. 

The fact that they *only* happen during the update cache process (seeing the stack trace, the new choke point moves to the ~partition_entry() destructor - haven't instrumented it yet) leads me to believe that this is not some external issue with preemptions as Tomek imagined, but the whole thing with the lower bound may be a red herring. I will try gathering more data.

**Comment by avikivity** [comment]
Profiles are not very useful for latency spikes, the average load drowns the spike.

**Comment by avikivity** [comment]
Try my latency tracer.

**Comment by glommer** [comment]
Ok, sorry: That was the last thing I have done yesterday, so probably a bit tired already. Looking at the cleanup code, it seems to me that it is just destroying all partitions in the memtable. So if we do this before the cache update, we will simply not insert anything - and that is why the problem doesn't show. It moves somewhere else, which at this point is very likely the same as #1551

**Comment by glommer** [comment]
@avikivity  The profile is from Tomek's request. The latency tracer is how I found out that the problem lies here in the first place. I have just instrumented the code manually so I would find out where exactly lies the problem inside the task.

**Comment by tgrabiec** [comment]
@avikivity The profile is an easy way to rule out a switchout as the cause. If we don't see switchouts during cache update, it would mean that they are not responsible for latency.

**Comment by glommer** [comment]
average number of comparisons = ~30

**Comment by avikivity** [comment]
That's exactly in the middle of my range estimate.  But what is the maximum number?  That's more interesting (not that I think we'll see surprises).

**Comment by avikivity** [comment]
Another alternative: lsa creates a large zone (many megabytes) and we see the kernel zeroing the pages.

**Comment by avikivity** [comment]
If that's correct, after a while the latency problems should disappear.

**Comment by tgrabiec** [comment]
For latencies in the order of 30ms that should easily show up on the perf profile. Maybe worth trying. One can zoom into `update_cache` as base of the flamegraph.

**Comment by glommer** [comment]
This is severely reduced when I forget about CPU0 and only look at other CPUs. This is a fairly modern baremetal hardware and /proc/interrupts doesn't show anything particularly concerning, except one interrupt that seems to be only being processed by CPU0 - ahci , which seems to be my disk controller.

I am setting smp_affinity to other cpus at /proc/irq but I don't see the irqs moving. When I capture those traces in cpus other than 0, I can still see update_cache contributing peaks, but that's nowhere near 30ms.

**Comment by glommer** [comment]
I didn't remember we had this bug opened. This was fixed by c32803f2f08060edc921812543c80d099c80596e
